New PR for MICE

kshedden opened the pull request:
Frank and I decided that I should resubmit a MICE PR from my account due to time limitations on his side.

This PR supercedes #1757.  It is a substantial overhaul of the code, but still contains all the main elements as designed and implemented by Frank in his GSOC project.

Here are two notebooks that illustrate how it can be used:

[Dead links, see below for updated working links]

http://nbviewer.ipython.org/urls/umich.box.com/shared/static/dkinhqyyz374k37tdwdi.ipynb

http://nbviewer.ipython.org/urls/umich.box.com/shared/static/pt8p8tctkl10t32xo97v.ipynb

Feedback welcome and encouraged.

josef-pkt commented:
just a quick look:
what happened to `get_distribution` attached to the existing models? Are they not needed anymore?

test coverage in this PR is considerably lower than in #1757, looks like just some smoke tests. Can the old tests be ported to this?

josef-pkt commented:
The notebooks and plots in it look very nice.

kshedden commented:
At some point we decided to pull back a bit and focus on predictive mean matching.  PMM seems to work quite well and apparently is the favored choice in other packages.  I thought it made sense to focus on other features like diagnostic plots rather than to continue wrestling with get_distribution.

I would be happy to get some feedback on the MICE implementation and can put time into upgrading the branch as needed.  

I will look into improving test coverage.

josef-pkt commented:
Hopefully @agartland can provide feedback. I can give a brief review after 0.7, but I have also several other PRs that I have kept waiting for too long.

@kshedden  Can you give a brief summary of the difference of this version and Franks #1757 version, especially with respect to what's in 1757 and not in here. Based on your last comment I assume that there are still things in 1757 that are not obsolete and might be ported to the new version at some point.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+  so the analysis model results are obtained using `next` or via any
+  standard python iterator pattern.
+
+* 'MICE' takes both a 'MICE_data' object and an analysis model
```
something like `MICE` is the wrapper class that does the full analysis, ....
IIUC, to get the contrast between the function that holds some pieces and the top class

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        self.cov = cov
+        self.cov_sqrt = np.linalg.cholesky(cov)
+
+    def rvs(self):
```
this is also in numpy.random.multivariate_normal  (which uses svd instead of cholesky)

josef-pkt replied:
but the numpy version won't store the sqrt, and redo it at each call

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+    Parameters
+    ----------
+    data : array-like object
+        The data set.
```
from the use of `dropna` below I assume this requires a dataframe not `array_like`

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        The data set.
+    perturbation_method : string
+        The default perturbation method
+    history_func : function
```
I guess a more standard terminology would be to call this a `call_back` such as in the optimization functions

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        # starting imputation.
+        self.data = self.data.fillna(self.data.mean())
+
+        self.k_pmm = 20
```
pure initialization might be  better with `None`, let the method handle the default?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        formula : string
+            Conditional formula for imputation. Defaults to a formula
+            with main effects for all other variables in dataset.  The
+            formula should only include an expression for the mean,
```
drop comma at end ?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+            with main effects for all other variables in dataset.  The
+            formula should only include an expression for the mean,
+            structure, e.g. use 'x1 + x2' not 'x4 ~ x1 + x2'.
+        model_class : statsmodels model
```
are "ducks" allowed? which kind of models?
(maybe becomes more obvious later)

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+            structure, e.g. use 'x1 + x2' not 'x4 ~ x1 + x2'.
+        model_class : statsmodels model
+            Conditional model for imputation. Defaults to OLS.
+        init_args : Dictionary
```
I prefer `init_kwds` as in our internal attribute, it's kwargs not args. ?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+            Conditional model for imputation. Defaults to OLS.
+        init_args : Dictionary
+            Keyword arguments passed to the model init method.
+        fit_args : Dictionary
```
same here `fit_kwds` ?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+            imp.data.to_csv('data%02d.csv' % j)
+"""
+
+class MICE_data(object):
```
general question (maybe obvious later)
Is there a MICE `method`, or currently it's only `pmm`, or it's not relevant here?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        ax.set_xticklabels(cols, rotation=90)
+
+        return fig
+
```
just in general: pep-8:  2 empty lines between methods and functions, IIRC, not yet enforced, but we try to get closer to it

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+                        plot_points=True, ax=None):
+        """
+        Create a scatterplot between the observed or imputed values of
+        a variable and the corresponding fitted values.
```
First line of docstring should be a **one**-line summary, otherwise it messes up sphinx doc rendering

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+                                            self.data,
+                                            **self.init_args[vname])
+        self.results[vname] = self.models[vname].fit(
+            **self.fit_args[vname])
```
question: Do we have to call this repeatedly for the same fit? in that case we would have redundant fit for generating rvs.
I guess, No, in this case where we will have new (imputed) `data` in each round.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+    def impute(self, vname):
+        # Wrap this in case we later add additional imputation
+        # methods.
+        self.impute_pmm(vname)
```
where would the imputation method be specified, when we add alternatives to `pmm`

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        self.perturb_params(vname)
+        self.impute(vname)
+
+    def impute_pmm(self, vname):
```
we don't use enough privatising of method, using leading underscore.
Are there method where it doesn't makes sense for the user to call those?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+
+        imputed_miss = np.array(endog_obs[iz])
+
+        self.store_changes(vname, imputed_miss)
```
general design question ( I don't have an overview yet)
Can we avoid that many methods work with sideeffects, i.e. changing attributes instead of returning?
I feel more comfortable (safer) if only the main function (like a fit method) changes state.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+        Number of imputation cycles to perform before fitting
+        the analysis models.
+    init_args : dict-like
+        Additional parameters for statsmodels model instance.
```
is this for the `analysis model`

`init_kwds`

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+    >>> plt.hist(params1)
+"""
+
+class MICE_model(object):
```
name is not pep-8  `MICEModel`? but is it really a `Model` class?
`MICEIterator`?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+    """ % {'mice_example_1' : _mice_example_1}
+
+    def __init__(self, model_formula, model_class, data,
+                 init_args={}, fit_args={}):
```
same here: `init_kwds`, `fit_kwds`
in general it's better to avoid mutable default arguments, i.e. use `init_kwds=None` and set 
```
if init_kwds = None:
    init_kwds = {}
```
(I don't remember the details, just being conservative (defensive programming).)

josef-pkt replied:
or the one-liner, something like
`self.init_kwds = init_kwds if init_kwds is not None else {}`

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1167 @@
+                                    init_args, fit_args)
+        self.model_chain = model_chain
+
+    def burnin(self, n_burnin):
```
private? `_burnin`

josef-pkt commented:
One reading through most parts:
I didn't see any serious problems (except low test coverage)

a few comments:

several naming questions `kwds`, class name, I didn't see much else

one interface question: I think MICE should get a `fit` method that combines `burnin` `run` and `combine` as a one-stop shopping method and closer to other models.

privatizing names with leading underscore: in general we don't use it enough in statsmodels. In this case (maybe because I don't know the structure so well but also because it has many methods across classes) I think there are methods where it doesn't make any sense for users to call them directly.

triggered partially by the previous comments: What do we do about other methods than `pmm`?
Do we have a rough idea about where we need to "future proof" this when we want to add more imputation methods?

josef-pkt commented:
semi random thought:
`start_params`: If we have models where fit is time consuming, then we should set start_params to use information from previous iterations to speed up the optimization.

josef-pkt commented:
question (One of my favorite topics, and a source of "surprises": the stale state. I have no idea yet.)

```
mi = mice.MICE("h39b ~ age + c28 + pcs + mcs37 + b05 + haartadhere", sm.OLS, imp)
mi.burnin(10)
mi.run(20, 5)
result = mi.combine()
mi.run(10, 5)
result2 = mi.combine()
```

What happens to `result`?
Do we need to warn, protect, ...?

josef-pkt commented:
aside `pmm` predictive mean matching and `psm` propensity score matching look very similar.

datnamer commented:
@josef-pkt  I'm in need of MICE and I would rather not dip into r. Is this in any sort of  rough usable state?

kshedden commented:
It works in all the examples we have tried, but admittedly it is not extensively tested.

Expect some renaming/refactoring later in the week when I get to addressing Josef's review.

datnamer commented:
@kshedden  thanks Ill give a run.

kshedden commented:
Updated notebooks:

https://umich.box.com/shared/static/5pcv0tbj302pxk7lpbalshlqddmmlk3m.ipynb
https://umich.box.com/shared/static/mv2wvdcicwl4ww1kamp8lkb7fahrnl7m.ipynb
https://umich.box.com/shared/static/yn2cttfg479wm2dmfnwnvopwbswgqkuy.ipynb

donaldbraman commented:
Just want to thank you for working on getting this up and running -- it's an crucial feature to serious analysis of incomplete data.

josef-pkt commented:
@kshedden  Can you rebase this? maybe in a few days, after fixing travis, and those in #2679
I'm planning another brief review and then merging it.

kshedden commented:
Yes, I will take care of this later in the week.

On Tue, Nov 3, 2015 at 10:15 PM, Josef Perktold <notifications@github.com>
wrote:

> @kshedden <https://github.com/kshedden> Can you rebase this? maybe in a
> few days, after fixing travis, and those in #2679
> <https://github.com/statsmodels/statsmodels/pull/2679>
> I'm planning another brief review and then merging it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/2076#issuecomment-153559391>
> .
>

datnamer commented:
Suhweet

josef-pkt commented:
TravisCI is working again, although we still have problems with latest numpy, 
newest pandas version is showing some convergence warnings with recent merges, and could possibly also have issues with this PR.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+            raise "cannot obtain predicted values from %s" % obj.__class__
+
+
+    def impute_pmm(self, vname):
```
we could add here a keyword argument for a selection/aggregation function as a callback.
It can default to the current random selection, but we could use non-uniform randomization or impute the mean of the neighbors for use in propensity score matching.
(not needed for this PR, it;s more for extensions and reusing code for other matching tasks)

josef-pkt commented:
Another general comment: the plots for checking the match between imputed and observed values look a bit similar to the diagnostics for treatment effect imputation that Stata added in version 14. (I didn't check details yet for that.)

This is still in the sandbox and we will have to move it soon.
I would like to put imputation together with `treatment`, but I'm still not sure `treatment` (used in my treatment effects PR) is a good name for both. Stata has now separate manuals for multiple imputation and treatment effects, but there is a large theoretical and code overlap.

josef-pkt commented:
TravisCI doesn't run any tests for this, I was checking coveralls and the mice.py module is missing. The number of tests run 4337 on python 3 is the same as in master

the names of the test classes have an unterline, which is not PEP8, but I thought nose would use them anyway. @kshedden  can you rename `Test_MICEData` -> `TestMICEData` and similar to see if it helps.

correction: that's not the reason, the test folder doesn't have an `__init__.py` so it won't be included as code, and it is not decleared as package data.
`statsmodels/sandbox/mice/tests/__init__.py`

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+    >>> j = 0
+    >>> for data in imp:
+            imp.data.to_csv('data%02d.csv' % j)
+            j += 1
```
the loop variable `data` is not used in the loop. Is this intentional? or should it be `data.to_csv('data%02d.csv' % j)`?

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+    Impute using default models, using the MICEData object as an
+    iterator.
+    %(_mice_data_example_2)s
+    Notes
```
doc build might break without blank line in front of section header

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+        self.ix_obs = {}
+        self.ix_miss = {}
+        for col in self.data.columns:
+            ix_obs, ix_miss = self._split_indices(self.data[col])
```
I usually prefer unique names for different things to avoid getting confused
e.g ix_obs_coll, ix_miss_coll   (or often I add an `_i` for one element

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+        def f(vec):
+            vec = vec.dropna()
+            mn = vec.mean()
+            ix = np.argmin(vec - mn)
```
shouldn't this be absolute value?
IIUC, it's the element closest to the mean, in abs norm.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+import numpy as np
+import patsy
+import statsmodels.api as sm
+import statsmodels
```
it's better to import from specific modules or subdirectories, than from `sm` api
I've only seen OLS so far.

josef-pkt replied:
avoids circular imports and doesn't load tsa or others if not needed.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+            self.conditional_formula[endog_name] = fml
+
+        if model_class is None:
+            self.model_class[endog_name] = sm.OLS
```
better to use from statsmodels.regression.linear_model, see comment at imports

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+
+        ix = self.ix_miss[col]
+        if len(ix) > 0:
+            self.data[col].iloc[ix] = vals
```
should this be a new indexing in pandas `data.ix[col, ix]` or loc or something, but with both slices together?

kshedden replied:
I am mixing label-based indexing (for columns) and position-based indexing (for rows).  'col' is a variable name and 'ix' is a numeric position.  I don't know how to do this kind of mixed indexing with a single call to ix or loc.

josef-pkt replied:
I'm just fighting with pandas over contingency table handling, trying to figure out at least some pivoting basics. `df_inj_count_other.ix[:2, 'All']`

ix handles both http://pandas.pydata.org/pandas-docs/stable/indexing.html

`.ix supports mixed integer and label based access. It is primarily label based, but will fall back to integer positional access unless the corresponding axis is of integer type`

AFAIR, we used this for the warnings that we assign to a copy in several places.

kshedden replied:
Would this create a problem if the row axis (which we have no control over) has integer type but is not identical to the positions?  The `ix_miss` values always refer to positions because they are obtained using `numpy.flatnonzero`.

My strategy was to internally use position-based indexing for rows (for performance), but we need to refer to variables (columns) by name because this is how the inputs are supplied.

Is there a downside to using `df[col_name].iloc[ii]`?  I don't see copy warnings on my setup.

josef-pkt replied:
then it's better to keep it the way you have it, I'm no expert on pandas.
Maybe the problem was only if rows where indexed first, which would make sense because columns are always homogeneous (same dtype) and stored in one array while rows are not.

josef-pkt commented:
Another general comment, maybe relevant later

The separate call to patsy dmatrices don't use the underlying state in stateful transforms or categorical encoding if one level is missing in a array. We might have to carry around the `design_info` for the full model if we create partial design matrices.
I was just reading through the patsy treatment of the kwds, and didn't see yet whether this uses full arrays or partial arrays, but with separate calls with observed and missing data there could be a problem. I don't expect a problem using full data, i.e. the current code has `self.data` in dmatrix, AFAICS.  
There might be some other limitation to whether the functions used in the formulas are in scope, but that's a special usage.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1287 @@
+
+        cov = self.results[vname].cov_params()
+        mu = self.results[vname].params
+        np.linalg.cholesky(cov) #!!!!!!!
```
what's this?  just a check for positive definite, and an exception if not?

josef-pkt commented:
random thoughts: two possible additions that might be useful if we want to continue from a stored dataset, for example,

- add a RandomState instance that users can set in `MICEData.__init__` 
- allow to start from an initial dataset, instead of default mean imputation.


- extra: we could `combine` and create Results instance from stored datasets. (sounds like what Stata is doing)

josef-pkt commented:
I'm stopping here for now. I still need to go over the results class and I skipped the plots.

Looks very good overall, and pretty much ready to merge.

I looked at the Stata manual again for multiple imputation, and there would be enough material to have it's own directory/subpackage, which would make it obsolete to come up with a name for imputation/treatment. 
I think it's better to move it in a new PR after merging this one, so github keeps displaying changes in this PR properly. (github doesn't follow renames and moves.)

josef-pkt commented:
missing test coverage looks mostly missing smoke tests for `hist` plot method (*) and `summary` method

(*) plot methods in models or results should have `plot_` prefix in name

I'll try to finish my review tomorrow.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+        self.data.update_all(n_burnin)
+
+        for j in range(n_imputations):
+            result = next(self)
```
Do we need that the class itself is an iterator?

When Skipper and I took over stats.models we removed all iterators from the classes and converted them to iteration loops inside methods. The main problem is that it is very difficult to keep track what is actually changed and where the side effects are of the iterator.

I haven't looked at a specific example yet, this is just general skepticism to possibly difficult to understand code.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+        scale_list = []
+        for results in self.results_list:
+            params_list.append(np.asarray(results.params))
+            cov_list.append(np.asarray(results.cov_params()))
```
We could save here the asarray detour if we use results._results, which is the results without wrapper.

(In cases where we could have wrapped or unwrapped results, we check it `results = getattr(results._results, results` might work)

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+        params = params_list.mean(0)
+
+        # The average of the within-imputation covariances
+        cov_within = sum(cov_list) / len(cov_list)
```
If we don't need `cov_list` otherwise, then we could accumulate `cov_within` inside the loop and save memory.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+
+        smry.add_dict(info, align='l', float_format=float_format)
+
+        param = summary2.summary_params(self, alpha=alpha)
```
Does this always use t-distribution for p-values and confidence interval?
I only see the table for the OLS examples in the notebooks.

josef-pkt commented:
possible bug in summary2 (bug in column label)  label is for `t` distribution, even if it's using normal?

Usecase: How do we handle repeated fitting during model search, evaluating different models? Do we have to start from scratch each time to get replicable random numbers given initial seed?
(extension to this for followup?)

josef-pkt commented:
I don't see test coverage for pvalues and confint. We have helper test functions to test consistency with normal or t distribution (given that we had some refactoring bugs with inherited pvalues and confint in the past.) 
Given the discussion last year during GSOC, we should also check that the correct df are used if the distribution is t. If there are ambiguities in which df to use, then our choice should be documented.

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+        for j in range(n_imputations):
+            result = next(self)
+            self.results_list.append(result)
+
```
Storing the entire results can become memory intensive.  I've done the same in adfuller, and user feedback was that the memory consumption goes up a lot with large arrays and many imputations/results. For the combine support we would only need to store params and scale and store or accumulate cov_params.
(storing full results could be optional or users could use a callback.)

josef-pkt commented on statsmodels/sandbox/mice/mice.py:
```diff
@@ -0,0 +1,1295 @@
+                                              self.data.data,
+                                              **self.init_kwds)
+        self.fit_kwds.update({"start_params": start_params})
+        result = model.fit(**self.fit_kwds)
```
another possible extension: 
`fit` from default start_params will be in many cases much slower than good start_params. It would be useful to get good start_parms from previous fits, however we don't have a standardized signature for start_params (not relevant in OLS and similar).
(We reuse start_params in some bootstrap methods but that's within a single model.)

josef-pkt commented:
Stata manual says they are using t distribution with missingness dependent degrees of freedom if number of imputations is small. (They mention normal distribution for large number of imputation, but without mentioning threshold and in a quick browsing I only see t distribution params tables.)

josef-pkt commented:
to summarize:

The plot methods need to be renamed to start with `plot_` for API consistency.

I'm not very happy about the iterator in MICE. (vague, because it might not be needed, and makes the class more complex than necessary.) If possible it would be better to change this before merging.

All the other issues can wait for a followup

- use of t distribution looks fine, but we need tests and possibly more code or review of inherited methods for other inference.
- restarting, storage, ..., are usability questions and can be changed over time.
- some code and memory streamlining could be now or later.

I haven't tried to understand all the details.

Overall I think the structure looks very good and flexible, and I expect that we will get many opportunities to build on and extend this. 

Can we add any of your notebooks to the documentation notebook examples?

kshedden commented:
I think I have addressed most of the critical issues.  When we move it out of the sandbox we can add additional features such as warm starts and saving the random state.

Updated notebooks are here:

https://umich.box.com/shared/static/mv2wvdcicwl4ww1kamp8lkb7fahrnl7m.ipynb
https://umich.box.com/shared/static/5pcv0tbj302pxk7lpbalshlqddmmlk3m.ipynb
https://umich.box.com/shared/static/yn2cttfg479wm2dmfnwnvopwbswgqkuy.ipynb

kshedden commented:
I set pdf_output = False and added a smoke test for summary.

josef-pkt commented:
Thanks @kshedden 

I'll merge when travis reports back

Do you have an opinion about directory name and location?

My initial thought was to use one folder for treatment and imputation, but now I think we just use two directories, and we can move this into `statsmodels/imputation` (in the next PR).

kshedden commented:
Thanks.  I don't have an opinion about directory location.

On Sat, Nov 14, 2015 at 1:44 PM, Josef Perktold <notifications@github.com>
wrote:

> Merged #2076 <https://github.com/statsmodels/statsmodels/pull/2076>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/2076#event-464583702>.
>

methodds commented:
Wow I am looking for a Python MICE implementation since several months, thank you so much for this work! Any idea when this will be included in statsmodels?

kshedden commented:
It's in.  If you have any questions or suggestions after using it email the
list:

https://groups.google.com/forum/#!forum/pystatsmodels

or open an issue on github.

On Sat, Nov 14, 2015 at 6:08 PM, Carsten Schwemmer <notifications@github.com
> wrote:

> Wow I am looking for a Python MICE implementation since several months,
> thank you so much for this work! Any idea when this will be included in
> statsmodels?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/2076#issuecomment-156755465>
> .
>

datnamer commented:
Nice!!!

donaldbraman commented:
THANK YOU!

josef-pkt commented:
Note: PR #3000  is moving mice to statsmodels.imputation.mice
There is currently no compatibility layer, for those already using it
